src/Aspire.Hosting.Azure/AzureProvisioningResource.cs

```diff
@@ -145,6 +145,82 @@ public static T CreateExistingOrNewProvisionableResource<T>(AzureResourceInfrast
         return provisionedResource;
     }
 
+    /// <summary>
+    /// Attempts to apply the name and (optionally) the resource group scope for <typeparamref name="TResource"/>
+    /// from an <see cref="ExistingAzureResourceAnnotation"/> attached to <paramref name="aspireResource"/>.
+    /// </summary>
+    /// <typeparam name="TResource">The concrete <see cref="ProvisionableResource"/> type.</typeparam>
+    /// <param name="aspireResource">The Aspire resource that may have an <see cref="ExistingAzureResourceAnnotation"/>.</param>
+    /// <param name="infra">The infrastructure used for converting parameters into provisioning expressions.</param>
+    /// <param name="resource">The provisionable resource to configure.</param>
+    /// <param name="setName">Delegate invoked to set the resolved name on <paramref name="resource"/>.</param>
+    /// <returns><see langword="true"/> if an <see cref="ExistingAzureResourceAnnotation"/> was present and applied; otherwise <see langword="false"/>.</returns>
+    /// <remarks>
+    /// When the annotation includes a resource group, a synthetic <c>scope</c> property is added to the resource's
+    /// provisionable properties to correctly scope the existing resource in the generated Bicep.
+    /// The caller is responsible for setting a generated name when the method returns <see langword="false"/>.
+    /// </remarks>
+    public static bool TryApplyExistingResourceNameAndScope<TResource>(IAzureResource aspireResource, AzureResourceInfrastructure infra, TResource resource, Action<TResource, BicepValue<string>> setName)
+        where TResource : ProvisionableResource
+    {
+        ArgumentNullException.ThrowIfNull(aspireResource);
+        ArgumentNullException.ThrowIfNull(infra);
+        ArgumentNullException.ThrowIfNull(resource);
+        ArgumentNullException.ThrowIfNull(setName);
+
+        if (!aspireResource.TryGetLastAnnotation<ExistingAzureResourceAnnotation>(out var existingAnnotation))
+        {
+            return false;
+        }
+
+        var existingResourceName = existingAnnotation.Name switch
+        {
+            ParameterResource nameParameter => nameParameter.AsProvisioningParameter(infra),
+            string s => new BicepValue<string>(s),
+            _ => throw new NotSupportedException($"Existing resource name type '{existingAnnotation.Name.GetType()}' is not supported.")
+        };
+
+        setName(resource, existingResourceName);
+
+        static bool ResourceGroupEquals(object existingResourceGroup, object? infraResourceGroup)
+        {
+            // We're in the resource group being created
+            if (infraResourceGroup is null)
+            {
+                return false;
+            }
+
+            // Compare the resource groups only if they are the same type (string or ParameterResource)
+            if (infraResourceGroup.GetType() == existingResourceGroup.GetType())
```

Comment:
Why this constraint? Couldn't we evaluate the value on a ParameterResource and compare it against a string resource group in a different annotation?

Reply:
That doesn't seem like a meaningful comparsion.